backend/utils/pubsub.js

```diff
@@ -0,0 +1,261 @@
+ * @returns {Object}
+ */
+export function createServer (httpServer: Object, options?: Object = {}): Object {
+  const server = new WebSocket.Server({
+    ...defaultOptions,
+    ...options,
+    server: httpServer
+  })
+  server.customClientHandlers = { ...options.clientHandlers }
+  server.customServerHandlers = { ...options.serverHandlers }
+  server.messageHandlers = { ...defaultMessageHandlers, ...options.messageHandlers }
+  server.pingInterval = undefined
+  server.subscribersByContractID = Object.create(null)
+
+  // Create and attach server-side WebSocket event listeners.
+  ;['close', 'connection', 'error', 'headers', 'listening'].forEach((name) => {
+    server.on(name, (...args) => {
+      console.log('[pubsub] Server event:', name)
+      try {
+        const customHandler = server.customServerHandlers['on' + name]
+        const defaultHandler = defaultServerHandlers['on' + name]
+        // Always call the default handler first.
+        if (defaultHandler) {
+          defaultHandler.call(server, ...args)
+        }
+        if (customHandler) {
+          customHandler.call(server, ...args)
+        }
+      } catch (error) {
+        logger(error)
+      }
+    })
+  })
+  // Setup a ping interval if required.
+  if (server.options.pingInterval > 0) {
+    server.pingInterval = setInterval(() => {
+      server.clients.forEach((client) => {
+        if (!client.isAlive) {
+          return client.terminate()
+        }
+        client.isAlive = false
+        client.ping()
+      })
+    }, server.options.pingInterval)
+  }
+  return server
+}
+
+const defaultOptions = {
+  clientTracking: true,
+  maxPayload: 6 * 1024 * 1024,
+  pingInterval: undefined
+}
+
+const generateSocketID = (() => {
+  let counter = 0
+
+  return () => counter++
+})()
+
+// ====== Server socket event handlers ====== //
+
+const defaultServerHandlers = Object.create(null)
+
+/**
+ * Emitted when a connection handshake completes.
+ *
+ * @param socket - The client socket that connected.
+ * @param request - The http GET request sent by the socket.
+ */
+defaultServerHandlers.onconnection = function onconnection (socket, request) {
+  socket.id = generateSocketID()
+  socket.isAlive = true
+  socket.server = this
+  socket.subscriptions = new Set()
+
+  const xForwardedFor = request.headers['x-forwarded-for']
+
+  const remoteAddress = (
+    typeof xForwardedFor === 'string' && xForwardedFor !== ''
+      ? xForwardedFor.split(/\s*,\s*/)[0]
+      : request.socket.remoteAddress
+  )
+  console.log(bold(`[pubsub] Socket ${socket.id} connected from ${remoteAddress}`))
+
+  // Create and attach socket event listeners.
+  ;['close', 'error', 'message', 'ping', 'pong'].forEach((eventName) => {
+    socket.on(eventName, (...args) => {
+      console.log(`[pubsub] Event '${eventName}' on socket ${socket.id}`, ...args)
+      const customHandler = socket.server.customClientHandlers['on' + eventName]
```

Comment:
We can also do as you suggest. I just thought it was nicer to pass options like `{ onerror: ..., onmessage: ... }` rather than  `{ error: ..., message: ... }`, because the common "on" prefix would act as a visual clue indicating that all these properties form a group of related event handlers.

As for the naming convention, it's still common to see such property names on objects implementing the `EventTarget` interface. For example `WebSocket` objects still have `onclose`, `onopen`, etc., although nowadays the preferred way to add listeners is via the `addEventListener()` method used with names without the "on" prefix.

Reply:
> We can also do as you suggest. I just thought it was nicer to pass options like `{ onerror: ..., onmessage: ... }` rather than `{ error: ..., message: ... }`, because the common "on" prefix would act as a visual clue indicating that all these properties form a group of related event handlers.

I would normally agree, but in this case it seems that all the handlers are explicitly defined in a property that has 'handlers' in the name, so I think that that would be sufficient visual cue to communicate this. Unless there's something I missed in how this is being used? I'm still getting used to the codebase :)

> As for the naming convention, it's still common to see such property names on objects implementing the `EventTarget` interface. For example `WebSocket` objects still have `onclose`, `onopen`, etc., although nowadays the preferred way to add listeners is via the `addEventListener()` method used with names without the "on" prefix.

Those property naming conventions are fairly widely considered to be a legacy holdover that everyone would rather forget about though :) For browser API spec people it makes sense to continue following it for consistency, but I don't think that we need to follow any such convention, especially if we're specifying them as options rather than setting them as properties anyway.